batch/batch/server/server.py

```diff
@@ -331,7 +331,8 @@ def from_record(record):
                        callback=record['callback'], userdata=userdata, user=record['user'],
                        always_run=record['always_run'], pvc_name=record['pvc_name'], pod_name=record['pod_name'],
                        exit_codes=exit_codes, duration=record['duration'], tasks=tasks,
-                       task_idx=record['task_idx'], state=record['state'], pvc_size=record['pvc_size'])
+                       task_idx=record['task_idx'], state=record['state'], pvc_size=record['pvc_size'],
+                       cancelled=record['cancelled'])
```

Comment:
I think you can change this line to also use the new record['cancelled'] property:

https://github.com/hail-is/hail/blob/e32d71ab54bf446db4958342538b1fdb968556dd/batch/batch/server/server.py#L685

Reply:
I think it should be the way it is because it's possible for a batch to have finished and then be cancelled. In that case, it should still show up as success.

Reply:
I thought about this a bit more. Is there a benefit to setting this state, rather than just storing n_cancelled? I think it may get confusing to have 2 records of state, with the same value (cancelled), but for different reasons (implicit, vs user specified).

I would define my api as you've done, and just have consuming functions understand that n_cancelled > 0 means the batch did not finish. That is the definition. It's in fact about the same amount of typing as checking whether batch.state != cancelled (batch.n_cancelled == 0).

Reply:
This stems from my desire to have a single source of truth. Here the source is n_cancelled. By going to state = " " we lose a bit of information, and make the api a bit more complex.

That being said, this is an opinion based on what I've seen, and you may have reasons that I do not understand for making this choice.

This comment does not need to be addressed to get this PR approved

Reply:
What if the number of jobs == 0? n_cancelled will be 0 in that case, but the batch should be set to cancelled. If I don't have a cancelled flag, then how will the first job know that the batch has been set to cancelled when it's time to set it's state in `create_if_ready`? The n_cancelled will be 0. I think the only way you can get rid of one of these states is to get rid of `n_cancelled` and change how batch state is computed. But I think we might want to know down the line how many jobs in the batch were cancelled vs succeeded vs failed.

Reply:
Then the batch will be set to cancelled because of Batch(cancelled=request[“cancelled”]). It just won’t be state=cancelled